sql/workload: add triggers to the random schema workload

Dedej-Bergin opened the pull request:
Added coverage for `CREATE TRIGGER` and `DROP TRIGGER` to the Random Schema Changer Workload.

Fixes: #146580
Epic: CRDB-42942
Release note: none

Dedej-Bergin committed:
sql/workload: add triggers to the random schema workload

Added coverage for `CREATE TRIGGER` and `DROP TRIGGER` to
the Random Schema Changer Workload.

Fixes: #146580
Epic: CRDB-42942
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147210)
<!-- Reviewable:end -->

spilchen reviewed (changes requested):
It's nice to see this being added to RSW

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-:-ORGhPYm3dQZ2-1_l5ij:b-epqjnj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek and @srosenberg)

___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORGZiRB49l5u4ogoTo5:bao8xi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
> ```Go
> 	// Create the function with legacy schema changer
> 	_, err = conn.Exec(ctx, `
> CREATE OR REPLACE FUNCTION log_change_timestamp()
> ```

If I understand the code correctly, this trigger function will be used in all triggers that are created? This is probably fine as a first step, but I think we should be generating a random trigger function. The reason is that most of the issues I have seen recently with triggers is because of dependencies in the trigger function. The random trigger function doesn't have to make any sense. It could be just running random select statements. That should be enough to establish more complex dependencies. I wonder if somehow we can call selectStmt or insertRow to generate the random SQL for the trigger function.
___
*[`pkg/workload/schemachange/operation_generator.go` line 4423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZ8WP-dBbx4-8ZpO6:-ORGZ8WP-dBbx4-8ZpO7:b-wwjys8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L4423)):*
> ```Go
> 	opStmt.potentialExecErrors.addAll(codesWithConditions{
> 		{pgcode.InvalidFunctionDefinition, hasFuncRefs},
> 		{pgcode.FeatureNotSupported, true},
> ```

What case was this added for?
___
*[`pkg/workload/schemachange/operation_generator.go` line 5369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgPKc7LGnKt-7q43R:-ORGgPKc7LGnKt-7q43S:b-s1qhrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5369)):*
> ```Go
> 	// Build the SQL statement
> 	var sqlStatement strings.Builder
> 	sqlStatement.WriteString(fmt.Sprintf("CREATE TRIGGER %s %s %s ON %s FOR EACH ROW EXECUTE FUNCTION log_change_timestamp()",
> ```

nit: I would use `fmt.Sprintf` directly and not `strings.Builder` since  you create the statement in a single line.
___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORGgmV73sEOzyfV1FBX:b-ft10g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
> ```Go
> 	// In some cases drop function will drop the trigger function.
> 	opStmt.potentialExecErrors.addAll(codesWithConditions{
> 		{code: pgcode.UndefinedFunction, condition: true},
> ```

Isn't this already check with `triggerFunctionExistts`?
___
*[`pkg/workload/schemachange/operation_generator.go` line 5402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGccR40kk6KUDjj5h0:-ORGccR40kk6KUDjj5h1:b2o8yi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5402)):*
> ```Go
> 		// Build the SQL statement
> 		var sqlStatement strings.Builder
> 		sqlStatement.WriteString(`DROP TRIGGER dummy_trigger ON dummy_table`)
> ```

nit: it seems overkill to use a string builder for this statement since it's constant
___
*[`pkg/workload/schemachange/operation_generator.go` line 5417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGcnAd0w_LQrwdK1_6:-ORGcnAd0w_LQrwdK1_7:b-k6k3le) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5417)):*
> ```Go
> 	// Build the SQL statement
> 	var sqlStatement strings.Builder
> 	sqlStatement.WriteString(fmt.Sprintf("DROP TRIGGER %s ON %s", triggerWithInfo.triggerName, &triggerWithInfo.table))
> ```

nit: is strings.Builder necessary here.?You generate the entire SQL statement in one WriteString call. So, you aren't taken advantage of the buffering that the builder gives you. I would suggest you get rid of strings.Builder and just store the statement in a regular string.
___
*[`pkg/workload/schemachange/operation_generator.go` line 5437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGbuJD9Ep5jkHhWC1J:-ORGbuJD9Ep5jkHhWC1K:bmy9xup) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5437)):*
> ```Go
> // It also returns a boolean indicating whether a trigger was found.
> func findExistingTrigger(
> 	ctx context.Context, tx pgx.Tx, og *operationGenerator,
> ```

nit: we don't use the `og` parameter
___
*[`pkg/workload/schemachange/operation_generator.go` line 5450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGb8r52Oi3DtFLr5Ck:-ORGb8r52Oi3DtFLr5Cl:b-rez8p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5450)):*
> ```Go
> 		FROM
> 			information_schema.triggers
> 		ORDER BY random()
> ```

nit: since we are using the `random()` function, we should call the `op.setSeedInDB` function.
___
*[`pkg/workload/schemachange/operation_generator.go` line 5470 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGaNNiBQrwF8BcUSVH:-ORGaNNiBQrwF8BcUSVI:bz3d58b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5470)):*
> ```Go
> 		triggerName: triggerName,
> 	}
> 	triggerExists = true
> ```

nit: I don't think we need the 2nd return parameter. Callers can just do a nil check on triggerWithInfo to know if a trigger exists. 


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORHBE5L0WcpXuFqTJKJ:b31lfn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Isn't this already check with `triggerFunctionExistts`?
</blockquote></details>

It is, but somehow we still hit this even outside the `triggerFunctionExists` condition.  I'm wondering if it's a timing issue, where by the time we execute the command, the query that we ran to check if the trigger function still exists is no longer correct.
___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORHBuP0BMYqvI3hkvE0:bfyysib) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

If I understand the code correctly, this trigger function will be used in all triggers that are created? This is probably fine as a first step, but I think we should be generating a random trigger function. The reason is that most of the issues I have seen recently with triggers is because of dependencies in the trigger function. The random trigger function doesn't have to make any sense. It could be just running random select statements. That should be enough to establish more complex dependencies. I wonder if somehow we can call selectStmt or insertRow to generate the random SQL for the trigger function.
</blockquote></details>

So one of the tricky things here is that creating the trigger function can only be done in the legacy schema changer, where as creating a trigger can only be done in the declarative schema changer, which is why I wasn't able to create a trigger function inside of the `createTrigger` method. With the current way where we only create one trigger function per 1 run of the whole test.  Would we want to make it where we still create only 1 trigger function per 1 run of the whole test, but just where we make that 1 function randomly generated for each whole test run?


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 4423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZ8WP-dBbx4-8ZpO6:-ORHH_WC9ymyRRatxTBn:b-kj9p79) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L4423)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

What case was this added for?
</blockquote></details>

I forgot to add a comment to explain this, I just added it now, thanks Matt!

```		// When create function calls a trigger function we get this error,
		// because trigger functions can only be called by triggers.
		{pgcode.FeatureNotSupported, true},
```

Without it we end up getting this error:
```
 "previousStatements": [
   "CREATE FUNCTION public.udf_w1_82()\n\tRETURNS VOID\n\tLANGUAGE SQL\n\tAS $$ SELECT NULL WHERE ((SELECT count(*) FROM public.table_w0_2 LIMIT 0) = 0) AND ('ouxc'::public.enum_w0_1 IS NULL) AND ('apm'::public.enum_w0_1 IS NULL) AND ((SELECT count(*) FROM public.table_w1_42 LIMIT 0) = 0) AND ((SELECT count(*) FROM public.trigger_log LIMIT 0) = 0) AND (SELECT public.log_change_timestamp() IS NOT NULL) AND ((SELECT count(*) FROM public.table_w0_20 LIMIT 0) = 0) $$"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:132: 
        	Error Trace:	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: trigger functions can only be called as triggers (SQLSTATE 0A000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:3322
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:546
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:634
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.7.darwin-arm64/src/runtime/asm_arm64.s:1223
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: trigger functions can only be called as triggers (SQLSTATE 0A000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
```
___
*[`pkg/workload/schemachange/operation_generator.go` line 5369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgPKc7LGnKt-7q43R:-ORHISBsD1yehOG8XSAF:b-4q6toa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5369)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: I would use `fmt.Sprintf` directly and not `strings.Builder` since  you create the statement in a single line.
</blockquote></details>

Sounds good, I went ahead and made this change, thanks Matt!
___
*[`pkg/workload/schemachange/operation_generator.go` line 5402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGccR40kk6KUDjj5h0:-ORHIkfV4adZBkVmuVNP:b-wn8ssl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5402)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: it seems overkill to use a string builder for this statement since it's constant
</blockquote></details>

Sounds good, I went ahead and simplified this, thanks Matt!
___
*[`pkg/workload/schemachange/operation_generator.go` line 5417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGcnAd0w_LQrwdK1_6:-ORHJLI5-M9LVfzeDoIb:b-4q6toa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5417)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: is strings.Builder necessary here.?You generate the entire SQL statement in one WriteString call. So, you aren't taken advantage of the buffering that the builder gives you. I would suggest you get rid of strings.Builder and just store the statement in a regular string.
</blockquote></details>

Sounds good, I went ahead and made this change, thanks Matt!
___
*[`pkg/workload/schemachange/operation_generator.go` line 5437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGbuJD9Ep5jkHhWC1J:-ORHJaYO0oE0E1eDj0yG:b-227li9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5437)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: we don't use the `og` parameter
</blockquote></details>

Sounds good, I went ahead and removed the `og` parameter, thanks Matt!


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/workload/schemachange/operation_generator.go` line 4423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZ8WP-dBbx4-8ZpO6:-ORHK_kk2Xo89rtXP8Nt:b-jw08dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L4423)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

I forgot to add a comment to explain this, I just added it now, thanks Matt!

```		// When create function calls a trigger function we get this error,
		// because trigger functions can only be called by triggers.
		{pgcode.FeatureNotSupported, true},
```

Without it we end up getting this error:
```
 "previousStatements": [
   "CREATE FUNCTION public.udf_w1_82()\n\tRETURNS VOID\n\tLANGUAGE SQL\n\tAS $$ SELECT NULL WHERE ((SELECT count(*) FROM public.table_w0_2 LIMIT 0) = 0) AND ('ouxc'::public.enum_w0_1 IS NULL) AND ('apm'::public.enum_w0_1 IS NULL) AND ((SELECT count(*) FROM public.table_w1_42 LIMIT 0) = 0) AND ((SELECT count(*) FROM public.trigger_log LIMIT 0) = 0) AND (SELECT public.log_change_timestamp() IS NOT NULL) AND ((SELECT count(*) FROM public.table_w0_20 LIMIT 0) = 0) $$"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:132: 
        	Error Trace:	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: trigger functions can only be called as triggers (SQLSTATE 0A000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:3322
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:546
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:634
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.7.darwin-arm64/src/runtime/asm_arm64.s:1223
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: trigger functions can only be called as triggers (SQLSTATE 0A000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
```
</blockquote></details>

This the comment I added, for some reason the first line was missing.

```
		// When create function calls a trigger function we get this error,
		// because trigger functions can only be called by triggers.
		{pgcode.FeatureNotSupported, true},
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin, @golgeek, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 4423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZ8WP-dBbx4-8ZpO6:-ORHtm2aAFxw7MxvcFVm:b-pccp7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L4423)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

This the comment I added, for some reason the first line was missing.

```
		// When create function calls a trigger function we get this error,
		// because trigger functions can only be called by triggers.
		{pgcode.FeatureNotSupported, true},
```
</blockquote></details>

I wonder if we can just omit that type when we collect all functions in the database (i.e. update `functionsQuery` in `createFunction`)
___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORHs0EQ5oQcA02n1yFu:b-9pp5bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

It is, but somehow we still hit this even outside the `triggerFunctionExists` condition.  I'm wondering if it's a timing issue, where by the time we execute the command, the query that we ran to check if the trigger function still exists is no longer correct.
</blockquote></details>

I thought the transaction semantics would prevent the timing scenario you mentioned. The transaction should retry if someone else dropped the function we were about to use. It's probably something we need to dig into to understand more.
___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORHrO-W0W15-DBxrPgD:bppatmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

So one of the tricky things here is that creating the trigger function can only be done in the legacy schema changer, where as creating a trigger can only be done in the declarative schema changer, which is why I wasn't able to create a trigger function inside of the `createTrigger` method. With the current way where we only create one trigger function per 1 run of the whole test.  Would we want to make it where we still create only 1 trigger function per 1 run of the whole test, but just where we make that 1 function randomly generated for each whole test run?
</blockquote></details>

I don't think we should have one trigger function for the entire run. It really limits the variability of the test. We won't get many interesting dependencies, which I think is a key goal of this work.

If we tie creating the trigger function with creating of the trigger, couldn't you set the `use_declarative_schema_changer` session variable temporarily to create the trigger function? 


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 5450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGb8r52Oi3DtFLr5Ck:-ORMHSAr46iKq3UDgN_7:bvywh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5450)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: since we are using the `random()` function, we should call the `op.setSeedInDB` function.
</blockquote></details>

Yes, just added this, thanks Matt!
___
*[`pkg/workload/schemachange/operation_generator.go` line 5470 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGaNNiBQrwF8BcUSVH:-ORMTcRT8NIRBiAVgdni:b-4q6toa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5470)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: I don't think we need the 2nd return parameter. Callers can just do a nil check on triggerWithInfo to know if a trigger exists.
</blockquote></details>

Sounds good, I went ahead and made this change, thanks Matt!
___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORMNrOZAuNtdHUgxab8:bquvwn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I don't think we should have one trigger function for the entire run. It really limits the variability of the test. We won't get many interesting dependencies, which I think is a key goal of this work.

If we tie creating the trigger function with creating of the trigger, couldn't you set the `use_declarative_schema_changer` session variable temporarily to create the trigger function?
</blockquote></details>

I tried to set and reset the `use_declarative_schema_changer` session variable, but I think since it runs async it doesn't take effect quickly enough and we end up getting the below error.

```
// createTrigger generates a CREATE TRIGGER statement.
func (og *operationGenerator) createTrigger(ctx context.Context, tx pgx.Tx) (*opStmt, error) {
	tableName, err := og.randTable(ctx, tx, og.pctExisting(true), "")
	if err != nil {
		return nil, err
	}

	tableExists, err := og.tableExists(ctx, tx, tableName)
	if err != nil {
		return nil, err
	}

		// First, ensure the trigger_log table exists
		_, err = tx.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS trigger_log (
			changed_at TIMESTAMP DEFAULT current_timestamp
		)`)
			if err != nil {
				return nil, err
			}

	// Create the trigger function if it doesn't exist
		// Remember the current setting
		var currentSetting string
		err = tx.QueryRow(ctx, "SHOW use_declarative_schema_changer").Scan(&currentSetting)
		if err != nil {
			return nil, err
		}

		// Set to use declarative schema changer
		_, err = tx.Exec(ctx, "SET use_declarative_schema_changer = on")
		if err != nil {
			return nil, err
		}

		// Create the function
		_, err = tx.Exec(ctx, `
CREATE OR REPLACE FUNCTION log_change_timestamp()
RETURNS TRIGGER AS $$
BEGIN
    INSERT INTO trigger_log
    VALUES (current_timestamp);
    RETURN NULL;
END;
$$ LANGUAGE PLpgSQL`)
		if err != nil {
			// Restore the original setting before returning error
			tx.Exec(ctx, fmt.Sprintf("SET use_declarative_schema_changer = %s", currentSetting))
			return nil, err
		}

		// Restore the original setting
		_, err = tx.Exec(ctx, fmt.Sprintf("SET use_declarative_schema_changer = %s", currentSetting))
		if err != nil {
			return nil, err
		}
```

Error when running workload with this implementation:

```
   schema_change_external_test.go:132: 
        	Error Trace:	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Failed to generate a random operation: failed generating operation: createTrigger: ERROR: *tree.CreateRoutine not implemented in the new schema changer (SQLSTATE XXUUU)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:537
        	            	  | [...repeated from below...]
        	            	Wraps: (4) ***UNEXPECTED ERROR; Failed to generate a random operation
        	            	Wraps: (5) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:219
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:524
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:634
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange_test.TestWorkload.func4.1
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.7.darwin-arm64/src/runtime/asm_arm64.s:1223
        	            	Wraps: (6) failed generating operation: createTrigger
        	            	Wraps: (7) ERROR: *tree.CreateRoutine not implemented in the new schema changer (SQLSTATE XXUUU)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *pgconn.PgError
        	Test:       	TestWorkload

```


<!-- Sent from Reviewable.io -->

spilchen reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin, @golgeek, and @srosenberg)

___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORMo4Ij8RrVxGJ2bX1I:bdz2rky) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

I tried to set and reset the `use_declarative_schema_changer` session variable, but I think since it runs async it doesn't take effect quickly enough and we end up getting the below error.

```
// createTrigger generates a CREATE TRIGGER statement.
func (og *operationGenerator) createTrigger(ctx context.Context, tx pgx.Tx) (*opStmt, error) {
	tableName, err := og.randTable(ctx, tx, og.pctExisting(true), "")
	if err != nil {
		return nil, err
	}

	tableExists, err := og.tableExists(ctx, tx, tableName)
	if err != nil {
		return nil, err
	}

		// First, ensure the trigger_log table exists
		_, err = tx.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS trigger_log (
			changed_at TIMESTAMP DEFAULT current_timestamp
		)`)
			if err != nil {
				return nil, err
			}

	// Create the trigger function if it doesn't exist
		// Remember the current setting
		var currentSetting string
		err = tx.QueryRow(ctx, "SHOW use_declarative_schema_changer").Scan(&currentSetting)
		if err != nil {
			return nil, err
		}

		// Set to use declarative schema changer
		_, err = tx.Exec(ctx, "SET use_declarative_schema_changer = on")
		if err != nil {
			return nil, err
		}

		// Create the function
		_, err = tx.Exec(ctx, `
CREATE OR REPLACE FUNCTION log_change_timestamp()
RETURNS TRIGGER AS $$
BEGIN
    INSERT INTO trigger_log
    VALUES (current_timestamp);
    RETURN NULL;
END;
$$ LANGUAGE PLpgSQL`)
		if err != nil {
			// Restore the original setting before returning error
			tx.Exec(ctx, fmt.Sprintf("SET use_declarative_schema_changer = %s", currentSetting))
			return nil, err
		}

		// Restore the original setting
		_, err = tx.Exec(ctx, fmt.Sprintf("SET use_declarative_schema_changer = %s", currentSetting))
		if err != nil {
			return nil, err
		}
```

Error when running workload with this implementation:

```
   schema_change_external_test.go:132: 
        	Error Trace:	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Failed to generate a random operation: failed generating operation: createTrigger: ERROR: *tree.CreateRoutine not implemented in the new schema changer (SQLSTATE XXUUU)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:537
        	            	  | [...repeated from below...]
        	            	Wraps: (4) ***UNEXPECTED ERROR; Failed to generate a random operation
        	            	Wraps: (5) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*operationGenerator).randOp
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:219
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:524
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:634
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange_test.TestWorkload.func4.1
        	            	  | 	/Users/bdedej/go/src/github.com/cockroachdb/cockroach/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	/Users/bdedej/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.7.darwin-arm64/src/runtime/asm_arm64.s:1223
        	            	Wraps: (6) failed generating operation: createTrigger
        	            	Wraps: (7) ERROR: *tree.CreateRoutine not implemented in the new schema changer (SQLSTATE XXUUU)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *pgconn.PgError
        	Test:       	TestWorkload

```
</blockquote></details>

You're setting a cluster wide variable. So, I wonder if you impacting other connections and that is where the error is coming from? Normally if you want to affect just a single connection you use the `SET SESSION...` syntax.

But it does look like CREATE FUNCTION is implemented in the DSC (`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go`). Some options aren't implemented, like the `REPLACE` option. But generally, it looks like you can create a function with the DSC. What I am missing?


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/schemachange.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZiRB49l5u4ogoTo4:-ORRYdss101R-ASHk1Aa:bru6mte) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/schemachange.go#L374)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

You're setting a cluster wide variable. So, I wonder if you impacting other connections and that is where the error is coming from? Normally if you want to affect just a single connection you use the `SET SESSION...` syntax.

But it does look like CREATE FUNCTION is implemented in the DSC (`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go`). Some options aren't implemented, like the `REPLACE` option. But generally, it looks like you can create a function with the DSC. What I am missing?
</blockquote></details>

Yes the `REPLACE` option was my issue, and I was able to make the trigger functions use selectStmt so that we can have random select statement, thanks Matt!


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 4423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGZ8WP-dBbx4-8ZpO6:-ORY-UMN7x6qglAmHZbl:bf2ntnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L4423)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I wonder if we can just omit that type when we collect all functions in the database (i.e. update `functionsQuery` in `createFunction`)
</blockquote></details>

Yes, I was able to omit it by updating functionsQuery, thanks Matt!
___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORY1NzxC_sm5kKTVhD4:b-rslhif) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I thought the transaction semantics would prevent the timing scenario you mentioned. The transaction should retry if someone else dropped the function we were about to use. It's probably something we need to dig into to understand more.
</blockquote></details>

Thanks Matt! This pgcode is no longer needed now that we create the trigger function inside of the `createTrigger` method, so I removed it.


<!-- Sent from Reviewable.io -->

spilchen reviewed (changes requested):
> Reviewed 1 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin, @golgeek, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORkPyrnDqxA926pet0W:bsamh4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Thanks Matt! This pgcode is no longer needed now that we create the trigger function inside of the `createTrigger` method, so I removed it.
</blockquote></details>

We already check `pgcode.UndefinedTable` in `opStmt.expectedExecError`. Why does this need to be handled as a potential error too?


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @golgeek, @spilchen, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORlzmu_CNk_X66hDerf:byabrnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

We already check `pgcode.UndefinedTable` in `opStmt.expectedExecError`. Why does this need to be handled as a potential error too?
</blockquote></details>

The check we do in `expectedExecError` is for the table we create the trigger on. But we also get this failure when the select statement in the trigger function tries to select a table that does not exist, and that is the one that the `potentialExecErrors` catches.

This is what the error looks like:
```
  "previousStatements": [
   "CREATE FUNCTION trigger_function_w1_38() RETURNS TRIGGER AS $$ BEGIN SELECT t0.col15_w1_16 AS col0 FROM public.table_w0_4 AS t0 ,public.table_w1_39 AS t1 FETCH FIRST 1 ROWS ONLY;RETURN NULL;END; $$ LANGUAGE PLpgSQL;CREATE TRIGGER trigger_w1_40 BEFORE INSERT ON public.table_w0_4 FOR EACH ROW EXECUTE FUNCTION trigger_function_w1_38()"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
Schema Workload Stats
Total Schema Statements Executed = 34
Total Schema Statements Succeeded = 34
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 23
Total Transactions Rolled Back = 18
Total Transactions Executed = 41
test logs left over in: /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
--- FAIL: TestWorkload (34.71s)
    test_log_scope.go:165: test logs captured to: /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_change_external_test.go:132: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: relation "public.table_w1_39" does not exist (SQLSTATE 42P01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3322
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:523
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:611
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1223
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: relation "public.table_w1_39" does not exist (SQLSTATE 42P01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
    schema_change_external_test.go:97: backup, tracing data, and system table dumps in /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
    panic.go:635: -- test log scope end --
FAIL
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I had another suggestion, but otherwise looks good to me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-:-ORm6OwyADYKpsuuKeSs:b-n57bkj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @Dedej-Bergin, @golgeek, and @srosenberg)

___
*[`pkg/workload/schemachange/operation_generator.go` line 5385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147210#-ORGgmV73sEOzyfV1FBW:-ORm69PG-X2rZjaRIgzf:b-s8hpp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db7f6b4dba499482a619988d2572e38c4dc2abc/pkg/workload/schemachange/operation_generator.go#L5385)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

The check we do in `expectedExecError` is for the table we create the trigger on. But we also get this failure when the select statement in the trigger function tries to select a table that does not exist, and that is the one that the `potentialExecErrors` catches.

This is what the error looks like:
```
  "previousStatements": [
   "CREATE FUNCTION trigger_function_w1_38() RETURNS TRIGGER AS $$ BEGIN SELECT t0.col15_w1_16 AS col0 FROM public.table_w0_4 AS t0 ,public.table_w1_39 AS t1 FETCH FIRST 1 ROWS ONLY;RETURN NULL;END; $$ LANGUAGE PLpgSQL;CREATE TRIGGER trigger_w1_40 BEFORE INSERT ON public.table_w0_4 FOR EACH ROW EXECUTE FUNCTION trigger_function_w1_38()"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
Schema Workload Stats
Total Schema Statements Executed = 34
Total Schema Statements Succeeded = 34
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 23
Total Transactions Rolled Back = 18
Total Transactions Executed = 41
test logs left over in: /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
--- FAIL: TestWorkload (34.71s)
    test_log_scope.go:165: test logs captured to: /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_change_external_test.go:132: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: relation "public.table_w1_39" does not exist (SQLSTATE 42P01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3322
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:523
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:611
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1223
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: relation "public.table_w1_39" does not exist (SQLSTATE 42P01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
    schema_change_external_test.go:97: backup, tracing data, and system table dumps in /Users/bdedej/GolandProjects/shickQuatro/cockroach/y/_tmp/9c9f43e6b40c9a0fcf6054db5e82708a/logTestWorkload1808917759
    panic.go:635: -- test log scope end --
FAIL
```
</blockquote></details>

Can we add a comment to the other `UndefinedTable` case (in the expected errors), to understand why we need two. It looks like because we have two we duplicated it by accident. But I see what you mean now.


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
bors r-

Dedej-Bergin commented:
TFTR!

bors r+
